### ISSUE TITLE
Fail fast on directory already exists

### ISSUE THREAD

**Issue by ssteinerx** [audreyr/cookiecutter#821]
* Cookiecutter version: 1.4.0
* Template project url: local
* Python version: 2.7
* Operating System: 10.10.6

### Description:

If the destination directory already exists (based on the project name), cookiecutter fails *after* all questions are asked.  

### What I've run:

cookiecutter, fill out name of existing directory should fail as soon as it can detect, not after answering all of the questions.

**Comment by hackebrot** [comment]
I see how checking the target directory while prompting would improve the user experience.

Do you want to create a PR, @ssteinerx?

**Comment by ssteinerx** [comment]
Unfortunately, `prompt.prompt_for_config` doesn't know anything about what is being asked; it just iterates through the given context and collects the specified information.

The 'context' is just the parsed `cookiecutter.json`, so there's no information there vis-à-vis validation of inputs.  The json just holds the prompts and default values.

Special-casing this one thing by name could be done but *ick*.

Without major rearrangement of `cookiecutter.json` to have validation hooks etc. (and essentially rewriting all the associated routines), or gluing on some special-case side-car validation gizmo onto `prompt_for_config`, there's no obvious/non-icky way of handling this with the current architecture.

**Comment by hackebrot** [comment]
Thanks for investigating, @ssteinerx! 🙇 

Sounds like another feature for #848 to me.

**Comment by ssteinerx** [comment]
@hackebrot Yes, I didn't have time to read the spec, but this type of fail-fast should be built in somehow.

**Comment by theodesp** [comment]
I see that we can decorate the `prompt_for_config` with something like a context aware prompt that will trigger extra actions when detecting special keys.

For example if it detects a key named `project_slug` will trigger a callback to check if the folder exists.

I think it's possible to do with out breaking changes. That will be a chance to refactor the `prompt.py`

**Comment by ssteinerx** [comment]
@theodesp I haven't looked at this in quite a while (like 8 months!), but I remember thinking that I'd want to have a way of indicating that `prompt` invokes `action-type` validator `action-type-validator` e.g.:

I haven't got time to look right this second, but I vaguely remember wanting something like:

prompt("Create the docs directory?", action_type="create-directory", validator="prompt-overwrite-if-exists-or-exit")

**Comment by theodesp** [comment]
Yes thats another good idea. As you said `prompt.prompt_for_config` doesn't know anything about what is being asked: So we need to keep the existing functionality. Maybe if we pass a map of keys to validators as a parameter that will get called if the particular key was prompted and filled as I dont see any support for validation in `click prompt` http://click.pocoo.org/5/api/#click.prompt
